src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,318 @@
+                self.leaves.remove(&index_value);
+            }
+
+            let sibling_node = index_value.sibling();
+            // node became a leaf only if it is a new node (it wasn't in nodes map)
+            if !self.nodes.contains_key(&sibling_node) {
+                self.leaves.insert(sibling_node);
+            }
+
+            // insert node from Merkle path to the nodes map
+            self.nodes.insert(sibling_node, hash.into());
+
+            Rpo256::merge(&index_value.build_node(node, hash.into()))
+        });
+
+        // if the path set is empty (the root is all ZEROs), set the root to the root of the added
+        // path; otherwise, the root of the added path must be identical to the current root
+        if self.root() == EMPTY_DIGEST {
+            self.nodes.insert(ROOT_INDEX, root);
+        } else if self.root() != root {
+            return Err(MerkleError::ConflictingRoots([*self.root(), *root].to_vec()));
+        }
+
+        // self.update_leaves()?;
+
+        Ok(())
+    }
+
+    /// Updates value of the leaf at the specified index returning the old leaf value.
+    ///
+    /// This also recomputes all hashes between the leaf and the root, updating the root itself.
+    pub fn update_leaf(
+        &mut self,
+        node_index: NodeIndex,
+        value: RpoDigest,
+    ) -> Result<RpoDigest, MerkleError> {
+        // check correctness of the depth and update it
+        Self::check_depth(node_index.depth())?;
+        self.update_depth(node_index.depth());
+
+        // insert NodeIndex to the leaves Set
+        self.leaves.insert(node_index);
+
+        // add node value to the nodes Map
+        let old_value = self.nodes.insert(node_index, value).unwrap_or(EMPTY_DIGEST);
+
+        // if the old value and new value are the same, there is nothing to update
+        if value == old_value {
+            return Ok(value);
+        }
+
+        let mut node_index = node_index;
+        let mut value = value;
+        for _ in 0..node_index.depth() {
+            let is_right = node_index.is_value_odd();
+            let (left, right) = if is_right {
+                let left_index = NodeIndex::new(node_index.depth(), node_index.value() - 1)?;
+                (
+                    self.nodes
+                        .get(&left_index)
+                        .cloned()
+                        .ok_or(MerkleError::NodeNotInSet(left_index))?,
+                    value,
+                )
+            } else {
+                let right_index = NodeIndex::new(node_index.depth(), node_index.value() + 1)?;
+                (
+                    value,
+                    self.nodes
+                        .get(&right_index)
+                        .cloned()
+                        .ok_or(MerkleError::NodeNotInSet(right_index))?,
+                )
+            };
+            node_index.move_up();
+            value = Rpo256::merge(&[left, right]);
+            self.nodes.insert(node_index, value);
```

Comment:
We should be able to simplify this quite a bit. Maybe something like:
```Rust
let sibling = self.nodes.get(&node_index.sibling()).expect("sibling should exist");
value = Rpo256::merge(&node_index.build_node(value, *sibling));
node_index.move_up();
self.nodes.insert(node_index, value);
```

Reply:
Ah, indeed! I always forget that we have a lot of convenient functions in the `NodeIndex` struct.